src/services/io/podio/JEventProcessorPODIO.cc

```diff
@@ -371,10 +371,13 @@ void JEventProcessorPODIO::FindCollectionsToWrite(const std::shared_ptr<const JE
 
 void JEventProcessorPODIO::Process(const std::shared_ptr<const JEvent> &event) {
 
-    std::lock_guard<std::mutex> lock(m_mutex);
+    std::unique_lock<std::mutex> lock(m_mutex, std::defer_lock);
+    lock.lock();
     if (m_is_first_event) {
         FindCollectionsToWrite(event);
+        m_is_first_event = false;
     }
+    lock.unlock();
```

Comment:
If I understand correctly (`m_is_first_event` monotonically changing from true to false), the lock can be in `if`, avoiding a lock/unlock after the first event. I'd want to avoid an unnecessary lock since it potentially introduces a wait for the write operation.

The (new) race condition to avoid then is `m_is_first_event` changing from true to false while the condition is evaluated, so we'd have to do a second locked `if` inside the unlocked `if`, but it would virtually never be used.

Reply:
You're describing the ABA problem. I think `std::call_once` is the tool for this purpose. (Just make sure that any code inside call_once is exception free because we've been bitten by bad standard library implementations before)

Reply:
> any code inside call_once is exception free

The code inside is an entire event processing in EICrecon :=)

Reply:
Explanation of double-checked locking and ABA problem for anyone who is interested: https://en.wikipedia.org/wiki/Double-checked_locking#Motivation_and_original_pattern

However, this is an optimization detail that I would avoid for now in favor of the simple, obvious solution. We'd like to verify:
- Consistent results in the output file
- A sensible cores vs throughput curve
- No data races uncovered by valgrind/tsan/whatever

Reply:
> > any code inside call_once is exception free
> 
> The code inside is an entire event processing in EICrecon :=)

Nope, just `FindCollectionsToWrite()` thankfully :)

Reply:
Ok, I see. At some point this ran through the whole first event to figure out which output names were activated, right? Or do I remember that wrong?